File: src/test/java/org/simbrain/util/projection/DatasetTest.java
package org.simbrain.util.projection;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class DatasetTest {

    Dataset data;

    @BeforeEach
    public void setUp() throws Exception {
        data = new Dataset(2);
        data.addPoint(new DataPoint(new double[]{1,0}));
    }

    @Test
    public void toleranceTest() {

        // Should be added. Within tolerance
        data.addPoint(new DataPoint(new double[]{2.1,0}), 1);

        assertEquals(2, data.getNumPoints());
        // Should not be added. Too close to previous.
        data.addPoint(new DataPoint(new double[]{1.5,0}), 1);
        assertEquals(2, data.getNumPoints());

    }

    @Test
    public void testLastPoint() {
        assertEquals(1.0, data.getLastAddedPoint().get(0),0 );
        assertEquals(0.0, data.getLastAddedPoint().get(1),0 );
    }

    // @Test
    // public void testND4JArray() {
    //     data = new Dataset(3);
    //     data.addPoint(new DataPoint(new double[]{1,0,1}));
    //     data.addPoint(new DataPoint(new double[]{1,1,1}));
    //     data.addPoint(new DataPoint(new double[]{0,0,1}));
    //     data.addPoint(new DataPoint(new double[]{-1,0,1}));
    //     INDArray arr = data.getArray();
    //     assertEquals(3, arr.columns());
    //     assertEquals(4, arr.rows());
    // }

    // TODO: Test isUnique
}

File: src/test/kotlin/org/simbrain/util/projection/DatasetTest.kt
package org.simbrain.util.projection

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class DatasetTest {

    var dataset: Dataset = Dataset(2)
    var p1 = DataPoint(doubleArrayOf(1.0, 0.0))
    var p2 = DataPoint(doubleArrayOf(0.5, 0.0))
    var p3 = DataPoint(doubleArrayOf(0.0, 0.0))

    @Test
    fun `test last added point`() {
        dataset.addPoint(p1)
        dataset.addPoint(p2)
        assertEquals(p2, dataset.lastAddedPoint)
    }

    @Test
    fun `test tolerance`() {
        dataset.addPoint(p1)
        dataset.addPoint(p2)
        // p3 is .5 away from the nearest point, within tolerance = 1, so don't add it
        dataset.addPoint(p3, 1.0)
        assertEquals(2, dataset.numPoints)
        // p3 is .5 away from the nearest point, abpve tolerance = .4, so add it
        dataset.addPoint(p3, .4)
        assertEquals(3, dataset.numPoints)
    }

    @Test
    fun `test distances`() {
        assertEquals(.5, dataset.getDistance(p1, p2))
        assertEquals(.5, dataset.getDistance(p2, p3))
        assertEquals(1.0, dataset.getDistance(p1, p3))
    }


}

